<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 4 Prepare training data | Genomic Prediction and Selection Manual</title>
<meta name="author" content="Marnin Wolfe">
<meta name="description" content="Context and Purpose: In this step, we do quality control, clean and format training data for further analysis. Upstream: Section 3 - training data download Downstream: pretty much everything...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 4 Prepare training data | Genomic Prediction and Selection Manual">
<meta property="og:type" content="book">
<meta property="og:url" content="https://wolfemd.github.io/GenomicSelectionManual/prepare-training-data.html">
<meta property="og:description" content="Context and Purpose: In this step, we do quality control, clean and format training data for further analysis. Upstream: Section 3 - training data download Downstream: pretty much everything...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 4 Prepare training data | Genomic Prediction and Selection Manual">
<meta name="twitter:description" content="Context and Purpose: In this step, we do quality control, clean and format training data for further analysis. Upstream: Section 3 - training data download Downstream: pretty much everything...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Genomic Prediction and Selection Manual</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Preamble</a></li>
<li><a class="" href="outline.html"><span class="header-section-number">2</span> Outline</a></li>
<li><a class="" href="download-training-data.html"><span class="header-section-number">3</span> Download training data</a></li>
<li><a class="active" href="prepare-training-data.html"><span class="header-section-number">4</span> Prepare training data</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/wolfemd/GenomicSelectionManual">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="prepare-training-data" class="section level1" number="4">
<h1>
<span class="header-section-number">4</span> Prepare training data<a class="anchor" aria-label="anchor" href="#prepare-training-data"><i class="fas fa-link"></i></a>
</h1>
<ul>
<li>
<strong>Context and Purpose:</strong> In this step, we do quality control, clean and format training data for further analysis.</li>
<li>
<strong>Upstream:</strong> Section <a href="download-training-data.html#download-training-data">3</a> - training data download</li>
<li>
<strong>Downstream:</strong> pretty much everything</li>
<li>
<strong>Inputs:</strong> “Raw” field trial data</li>
<li>
<strong>Expected outputs:</strong> “Cleaned” field trial data</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span><span class="op">)</span>; 
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/genomicMateSelectR">genomicMateSelectR</a></span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="read-db-data" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Read DB data<a class="anchor" aria-label="anchor" href="#read-db-data"><i class="fas fa-link"></i></a>
</h2>
<p>Load the phenotype and metadata downloads into R.</p>
<p>I built a function <code>readDBdata</code> that simply wraps around <code>read.csv</code>, reads and merges the metadata to the plot-basis data. The <code>metadataFile=</code> argument can be left NULL.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/readDBdata.html">readDBdata</a></span><span class="op">(</span>phenotypeFile <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data"</span>,<span class="st">"phenotype.csv"</span><span class="op">)</span>,
                   metadataFile <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data"</span>,<span class="st">"metadata.csv"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = c("studyYear", "programDbId", "programName", "programDescription", "studyDbId", "studyName", "studyDescription", "studyDesign", "plotWidth", "plotLength", "fieldSize", "fieldTrialIsPlannedToBeGenotyped", "fieldTrialIsPlannedToCross", "plantingDate", "harvestDate", "locationDbId", "locationName")</span></code></pre></div>
<p><strong>HINT:</strong> At any point in the manual, if I reference or use a custom function in the <code>genomicMateSelectR</code>, I encourage you to check out the reference page for that function, e.g. <a href="https://wolfemd.github.io/genomicMateSelectR/reference/readDBdata.html"><code>readDBdata()</code></a>. Or look at the code yourself by typing e.g. <code>readDBdata</code> at the R console or heading to the GitHub repo.</p>
</div>
<div id="group-and-select-trials-to-analyze" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> Group and select trials to analyze<a class="anchor" aria-label="anchor" href="#group-and-select-trials-to-analyze"><i class="fas fa-link"></i></a>
</h2>
<p>This step is present in my standard pipeline because I often bulk-download data for a breeding program and then sort through the trials-of-interest after the fact.</p>
<p>I think the better future way is to use trial-lists on BreedBase to communicate exactly which trials, years, locations are desired at download, maybe also which traits.</p>
<p>Unecessary for the example dataset chosen.</p>
</div>
<div id="traits-and-trait-abbreviations" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Traits and Trait Abbreviations<a class="anchor" aria-label="anchor" href="#traits-and-trait-abbreviations"><i class="fas fa-link"></i></a>
</h2>
<p>Cassavabase downloads use very long column-names corresponding to the <a href="https://cropontology.org/ontology/CO_334">full trait-ontology name</a>. For convenience, I replace these names with abbreviations, documented here. For eventual upload of analysis results, names will need to be restored to ontology terms.</p>
<p>I also use this opportunity to subselect traits.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traitabbrevs</span><span class="op">&lt;-</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span><span class="op">(</span><span class="op">~</span><span class="va">TraitAbbrev</span>,<span class="op">~</span><span class="va">TraitName</span>,
        <span class="st">"CMD1S"</span>,<span class="st">"cassava.mosaic.disease.severity.1.month.evaluation.CO_334.0000191"</span>,
        <span class="st">"CMD3S"</span>,<span class="st">"cassava.mosaic.disease.severity.3.month.evaluation.CO_334.0000192"</span>,
        <span class="st">"CMD6S"</span>,<span class="st">"cassava.mosaic.disease.severity.6.month.evaluation.CO_334.0000194"</span>,
        <span class="st">"DM"</span>,<span class="st">"dry.matter.content.percentage.CO_334.0000092"</span>,
        <span class="st">"RTWT"</span>,<span class="st">"fresh.storage.root.weight.per.plot.CO_334.0000012"</span>,
        <span class="st">"NOHAV"</span>,<span class="st">"plant.stands.harvested.counting.CO_334.0000010"</span><span class="op">)</span>
<span class="va">traitabbrevs</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu">gt</span><span class="fu">::</span><span class="fu"><a href="https://gt.rstudio.com/reference/gt.html">gt</a></span><span class="op">(</span><span class="op">)</span><span class="co">#rmarkdown::paged_table()</span></code></pre></div>
<div id="csyvgbtwtr" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#csyvgbtwtr .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#csyvgbtwtr .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#csyvgbtwtr .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#csyvgbtwtr .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#csyvgbtwtr .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#csyvgbtwtr .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#csyvgbtwtr .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#csyvgbtwtr .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#csyvgbtwtr .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#csyvgbtwtr .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#csyvgbtwtr .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#csyvgbtwtr .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#csyvgbtwtr .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#csyvgbtwtr .gt_from_md > :first-child {
  margin-top: 0;
}

#csyvgbtwtr .gt_from_md > :last-child {
  margin-bottom: 0;
}

#csyvgbtwtr .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#csyvgbtwtr .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#csyvgbtwtr .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#csyvgbtwtr .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#csyvgbtwtr .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#csyvgbtwtr .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#csyvgbtwtr .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#csyvgbtwtr .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#csyvgbtwtr .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#csyvgbtwtr .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#csyvgbtwtr .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#csyvgbtwtr .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#csyvgbtwtr .gt_left {
  text-align: left;
}

#csyvgbtwtr .gt_center {
  text-align: center;
}

#csyvgbtwtr .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#csyvgbtwtr .gt_font_normal {
  font-weight: normal;
}

#csyvgbtwtr .gt_font_bold {
  font-weight: bold;
}

#csyvgbtwtr .gt_font_italic {
  font-style: italic;
}

#csyvgbtwtr .gt_super {
  font-size: 65%;
}

#csyvgbtwtr .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">TraitAbbrev</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">TraitName</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">CMD1S</td>
<td class="gt_row gt_left">cassava.mosaic.disease.severity.1.month.evaluation.CO_334.0000191</td>
</tr>
<tr>
<td class="gt_row gt_left">CMD3S</td>
<td class="gt_row gt_left">cassava.mosaic.disease.severity.3.month.evaluation.CO_334.0000192</td>
</tr>
<tr>
<td class="gt_row gt_left">CMD6S</td>
<td class="gt_row gt_left">cassava.mosaic.disease.severity.6.month.evaluation.CO_334.0000194</td>
</tr>
<tr>
<td class="gt_row gt_left">DM</td>
<td class="gt_row gt_left">dry.matter.content.percentage.CO_334.0000092</td>
</tr>
<tr>
<td class="gt_row gt_left">RTWT</td>
<td class="gt_row gt_left">fresh.storage.root.weight.per.plot.CO_334.0000012</td>
</tr>
<tr>
<td class="gt_row gt_left">NOHAV</td>
<td class="gt_row gt_left">plant.stands.harvested.counting.CO_334.0000010</td>
</tr>
</tbody>
</table></div>
</div>
<p>Run function <code><a href="https://rdrr.io/pkg/genomicMateSelectR/man/renameAndSelectCols.html">renameAndSelectCols()</a></code> to rename columns and remove unselected traits.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/renameAndSelectCols.html">renameAndSelectCols</a></span><span class="op">(</span><span class="va">traitabbrevs</span>,
                            indata<span class="op">=</span><span class="va">dbdata</span>,
                            customColsToKeep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"observationUnitName"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "TraitName"</span></code></pre></div>
</div>
<div id="qc-trait-values" class="section level2" number="4.4">
<h2>
<span class="header-section-number">4.4</span> QC Trait Values<a class="anchor" aria-label="anchor" href="#qc-trait-values"><i class="fas fa-link"></i></a>
</h2>
<p>At this point in the pipeline, we should check the all trait values are in allowable ranges.
Different ways to approach this. Feel free to make some plots of your data!</p>
<p>The database also has mechanisms to ensure trait values are only within allowable ranges.
Nevertheless, as a habit, I have an simple <em>ad hoc</em> approach to this:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># comment out the traits not present in this dataset</span>
<span class="va">dbdata</span><span class="op">&lt;-</span><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>CMD1S<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">CMD1S</span><span class="op">&lt;</span><span class="fl">1</span> <span class="op">|</span> <span class="va">CMD1S</span><span class="op">&gt;</span><span class="fl">5</span>,<span class="cn">NA</span>,<span class="va">CMD1S</span><span class="op">)</span>,
         CMD3S<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">CMD3S</span><span class="op">&lt;</span><span class="fl">1</span> <span class="op">|</span> <span class="va">CMD3S</span><span class="op">&gt;</span><span class="fl">5</span>,<span class="cn">NA</span>,<span class="va">CMD3S</span><span class="op">)</span>,
         <span class="co">#CMD6S=ifelse(CMD6S&lt;1 | CMD6S&gt;5,NA,CMD6S), </span>
         <span class="co">#CMD9S=ifelse(CMD9S&lt;1 | CMD9S&gt;5,NA,CMD9S),</span>
         <span class="co"># CGM=ifelse(CGM&lt;1 | CGM&gt;5,NA,CGM),</span>
         <span class="co"># CGMS1=ifelse(CGMS1&lt;1 | CGMS1&gt;5,NA,CGMS1),</span>
         <span class="co"># CGMS2=ifelse(CGMS2&lt;1 | CGMS2&gt;5,NA,CGMS2),</span>
         DM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">DM</span><span class="op">&gt;</span><span class="fl">100</span> <span class="op">|</span> <span class="va">DM</span><span class="op">&lt;=</span><span class="fl">0</span>,<span class="cn">NA</span>,<span class="va">DM</span><span class="op">)</span>,
         RTWT<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">RTWT</span><span class="op">==</span><span class="fl">0</span> <span class="op">|</span> <span class="va">NOHAV</span><span class="op">==</span><span class="fl">0</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">NOHAV</span><span class="op">)</span>,<span class="cn">NA</span>,<span class="va">RTWT</span><span class="op">)</span>,
         <span class="co"># SHTWT=ifelse(SHTWT==0 | NOHAV==0 | is.na(NOHAV),NA,SHTWT),</span>
         <span class="co"># RTNO=ifelse(RTNO==0 | NOHAV==0 | is.na(NOHAV),NA,RTNO),</span>
         NOHAV<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">NOHAV</span><span class="op">==</span><span class="fl">0</span>,<span class="cn">NA</span>,<span class="va">NOHAV</span><span class="op">)</span>,
         NOHAV<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">NOHAV</span><span class="op">&gt;</span><span class="fl">42</span>,<span class="cn">NA</span>,<span class="va">NOHAV</span><span class="op">)</span>
         <span class="co">#RTNO=ifelse(!RTNO %in% 1:10000,NA,RTNO)</span>
         <span class="op">)</span></code></pre></div>
</div>
<div id="post-qc-composite-traits" class="section level2" number="4.5">
<h2>
<span class="header-section-number">4.5</span> Post-QC: composite traits<a class="anchor" aria-label="anchor" href="#post-qc-composite-traits"><i class="fas fa-link"></i></a>
</h2>
<p>Now that component traits are QC’d, it’s time to compute any composite traits.</p>
<p>By composite traits, I mean traits computed from combinations of other traits.</p>
<p>Examples for cassava: season-wide mean disease severity, harvest index, and fresh root yield.</p>
</div>
<div id="season-wide-mean-disease-severity" class="section level2" number="4.6">
<h2>
<span class="header-section-number">4.6</span> Season-wide mean disease severity<a class="anchor" aria-label="anchor" href="#season-wide-mean-disease-severity"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># [NEW AS OF APRIL 2021]</span>
<span class="co">## VERSION with vs. without CBSD</span>
<span class="co">## Impervious to particular timepoints between 1, 3, 6 and 9 scores</span>

<span class="co"># Without CBSD (West Africa)</span>
<span class="va">dbdata</span><span class="op">&lt;-</span><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>MCMDS<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">.</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CMD1S"</span>,<span class="st">"CMD3S"</span>,<span class="st">"CMD6S"</span>,<span class="st">"CMD9S"</span><span class="op">)</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">any_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CMD1S"</span>,<span class="st">"CMD3S"</span>,<span class="st">"CMD6S"</span>,<span class="st">"CMD9S"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># With CBSD (East Africa)</span>
<span class="co"># dbdata&lt;-dbdata %&gt;% </span>
<span class="co">#   mutate(MCMDS=rowMeans(.[,colnames(.) %in% c("CMD1S","CMD3S","CMD6S","CMD9S")], na.rm = T),</span>
<span class="co">#          MCBSDS=rowMeans(.[,colnames(.) %in% c("CBSD1S","CBSD3S","CBSD6S","CBSD9S")], na.rm = T)) %&gt;% </span>
<span class="co">#   select(-any_of(c("CMD1S","CMD3S","CMD6S","CMD9S","CBSD1S","CBSD3S","CBSD6S","CBSD9S")))</span></code></pre></div>
</div>
<div id="fresh-root-yield-fyld" class="section level2" number="4.7">
<h2>
<span class="header-section-number">4.7</span> Fresh root yield (FYLD)<a class="anchor" aria-label="anchor" href="#fresh-root-yield-fyld"><i class="fas fa-link"></i></a>
</h2>
<p><strong>RTWT</strong> (fresh root weight per plot in kg) –&gt; <strong>FYLD</strong> (fresh root yield in tons per hectare)</p>
<p><span class="math display">\[FYLD = \frac{RTWT_{kg / plot}}{MaxHarvestedPlantsPerPlot \times PlantSpacing}\times10 \]</span></p>
<p><strong>NOTE:</strong> <em>MaxHarvestedPlantsPerPlot</em> in formula above is to distinguish from the <em>plantsPerPlot</em> meta-data field, in case that a net-plot harvest is used. In other words, the value should be the total number of plants intended for harvest in a plot.</p>
<p><em>PlantSpacing</em> is the area in <span class="math inline">\(m^2\)</span> per plant.</p>
<p>In the example trial data, we the <code>plantsPerPlot</code> meta-data field is empty. Luckily, since there are only two trials, we make a quick summary of the NOHAV data, to determine the correct values.</p>
<p><strong>RECOMMEND INPUTING plantsPerPlot meta-data to cassavabase</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">studyYear</span>,<span class="va">studyName</span>,<span class="va">studyDesign</span>,<span class="va">plotWidth</span>,<span class="va">plotLength</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 9 × 6</span>
<span class="co">#&gt;   studyYear studyName studyDesign plotWidth plotLength     n</span>
<span class="co">#&gt;       &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;           &lt;int&gt;      &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1      2019 19.GS.C1… Alpha               4        4     125</span>
<span class="co">#&gt; 2      2019 19.GS.C2… Alpha               4        4      68</span>
<span class="co">#&gt; 3      2019 19.GS.C2… Alpha               4        4      72</span>
<span class="co">#&gt; 4      2019 19.GS.C2… RCBD                4        4      66</span>
<span class="co">#&gt; 5      2019 19.GS.C2… Alpha               4        4      72</span>
<span class="co">#&gt; 6      2019 19.GS.C4… Alpha               2        4     270</span>
<span class="co">#&gt; 7      2019 19.GS.C4… Alpha               3        2.5   273</span>
<span class="co">#&gt; 8      2019 19.GS.C4… RCBD                1        2.5   777</span>
<span class="co">#&gt; 9      2019 19geneti… Augmented           1        2.5   810</span></code></pre></div>
<p>So the GS.C1 trial has 2.5 <span class="math inline">\(m^2\)</span> plots, the GeneticGain trial has 8 <span class="math inline">\(m^2\)</span>.</p>
<p>A quick density plot reveals that the GeneticGain trial was likely planted with 10 plants/plot, and the GS.C1.CE 5 plants/plot. (<em>Disclosure:</em> I know this is true from experience.)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">NOHAV</span>, fill<span class="op">=</span><span class="va">studyName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 921 rows containing non-finite values</span>
<span class="co">#&gt; (stat_density).</span></code></pre></div>
<div class="inline-figure"><img src="02-training_data_QC_files/figure-html/unnamed-chunk-8-1.png" width="672"></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="co"># plot area in meters squared</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>plotArea<span class="op">=</span><span class="va">plotWidth</span><span class="op">*</span><span class="va">plotLength</span>,
            <span class="co"># Number of plants per plot</span>
            plantsPerPlot<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">studyName</span><span class="op">==</span><span class="st">"13geneticgainIB"</span>,<span class="fl">10</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>PlantSpacing<span class="op">=</span><span class="va">plotArea</span><span class="op">/</span><span class="va">plantsPerPlot</span>,
            FYLD<span class="op">=</span><span class="va">RTWT</span><span class="op">/</span><span class="op">(</span><span class="va">plantsPerPlot</span><span class="op">*</span><span class="va">PlantSpacing</span><span class="op">)</span><span class="op">*</span><span class="fl">10</span><span class="op">)</span>
<span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">FYLD</span>,fill<span class="op">=</span><span class="va">studyName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 943 rows containing non-finite values</span>
<span class="co">#&gt; (stat_density).</span></code></pre></div>
<div class="inline-figure"><img src="02-training_data_QC_files/figure-html/unnamed-chunk-10-1.png" width="672"></div>
<p>Additional things to compute:</p>
<ol style="list-style-type: decimal">
<li>log-transform yield traits: this is a habit based on experience. Linear mixed-models should have normally distributed homoskedastic residuals, if they don’t log-transform the response variable often helps. For FYLD and related traits, I always log-transform.</li>
</ol>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># I log transform yield traits </span>
<span class="co"># to satisfy homoskedastic residuals assumption </span>
<span class="co"># of linear mixed models</span>
<span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>DYLD<span class="op">=</span><span class="va">FYLD</span><span class="op">*</span><span class="op">(</span><span class="va">DM</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span>,
            logFYLD<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">FYLD</span><span class="op">)</span>,
            logDYLD<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">DYLD</span><span class="op">)</span>,
            PropNOHAV<span class="op">=</span><span class="va">NOHAV</span><span class="op">/</span><span class="va">plantsPerPlot</span><span class="op">)</span> 
<span class="co"># remove non transformed / per-plot (instead of per area) traits</span>
<span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">RTWT</span>,<span class="op">-</span><span class="va">FYLD</span>,<span class="op">-</span><span class="va">DYLD</span><span class="op">)</span>
<span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">logFYLD</span>,fill<span class="op">=</span><span class="va">studyName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 943 rows containing non-finite values</span>
<span class="co">#&gt; (stat_density).</span></code></pre></div>
<div class="inline-figure"><img src="02-training_data_QC_files/figure-html/unnamed-chunk-11-1.png" width="672"></div>
<p>Debatable whether this is better. Let’s not dwell on it. Onward!</p>
</div>
<div id="check-genotype-to-phenotype-matches" class="section level2" number="4.8">
<h2>
<span class="header-section-number">4.8</span> Check genotype-to-phenotype matches<a class="anchor" aria-label="anchor" href="#check-genotype-to-phenotype-matches"><i class="fas fa-link"></i></a>
</h2>
<p>In this step, in the past (see <a href="https://wolfemd.github.io/IITA_2021GS/01-cleanTPdata.html#%5BUser_input%5D_Assign_genos_to_phenos">here</a> for an example), I have had to use multiple external flat files and custom code, built over time, to figure out which DNA samples, if any, match eatch field plot.</p>
<p>With both genotypes and phenotypes derived from the database, this should not be an issue for this tutorial.</p>
</div>
<div id="check-detect-experimental-designs" class="section level2" number="4.9">
<h2>
<span class="header-section-number">4.9</span> Check / “detect” experimental designs?<a class="anchor" aria-label="anchor" href="#check-detect-experimental-designs"><i class="fas fa-link"></i></a>
</h2>
<p>In this step, in the past, I have not been certain of the experimental designs of the trials I had downloaded. I was also not certain how the designs were represented in the column-names. For this reason, I developed another <em>ad hoc</em> custom code to “detect” the designs. I built the <code>genomicMateSelectR</code> function <code><a href="https://rdrr.io/pkg/genomicMateSelectR/man/detectExptDesigns.html">detectExptDesigns()</a></code>. See an example <a href="https://wolfemd.github.io/IITA_2021GS/01-cleanTPdata.html#Detect_experimental_designs">here</a>.</p>
<p>Should not be necessary for the example dataset. However, let’s check:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dbdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">studyName</span>,<span class="va">studyDesign</span>, <span class="va">numberBlocks</span>,<span class="va">replicate</span>, <span class="va">blockNumber</span>, <span class="va">entryType</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 224 × 7</span>
<span class="co">#&gt;    studyName  studyDesign numberBlocks replicate blockNumber</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;              &lt;int&gt;     &lt;int&gt;       &lt;int&gt;</span>
<span class="co">#&gt;  1 19.GS.C1.… Alpha                 NA         1           1</span>
<span class="co">#&gt;  2 19.GS.C1.… Alpha                 NA         1           1</span>
<span class="co">#&gt;  3 19.GS.C1.… Alpha                 NA         2           2</span>
<span class="co">#&gt;  4 19.GS.C1.… Alpha                 NA         2           2</span>
<span class="co">#&gt;  5 19.GS.C1.… Alpha                 NA         3           3</span>
<span class="co">#&gt;  6 19.GS.C1.… Alpha                 NA         3           3</span>
<span class="co">#&gt;  7 19.GS.C2.… Alpha                  6         1           1</span>
<span class="co">#&gt;  8 19.GS.C2.… Alpha                  6         1           1</span>
<span class="co">#&gt;  9 19.GS.C2.… Alpha                  6         1           2</span>
<span class="co">#&gt; 10 19.GS.C2.… Alpha                  6         1           2</span>
<span class="co">#&gt; # … with 214 more rows, and 2 more variables:</span>
<span class="co">#&gt; #   entryType &lt;chr&gt;, n &lt;int&gt;</span></code></pre></div>
</div>
<div id="checklist-and-kpi" class="section level2" number="4.10">
<h2>
<span class="header-section-number">4.10</span> Checklist and KPI<a class="anchor" aria-label="anchor" href="#checklist-and-kpi"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>[ ]</li>
</ul>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="download-training-data.html"><span class="header-section-number">3</span> Download training data</a></div>
<div class="next"><a href="references.html">References</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#prepare-training-data"><span class="header-section-number">4</span> Prepare training data</a></li>
<li><a class="nav-link" href="#read-db-data"><span class="header-section-number">4.1</span> Read DB data</a></li>
<li><a class="nav-link" href="#group-and-select-trials-to-analyze"><span class="header-section-number">4.2</span> Group and select trials to analyze</a></li>
<li><a class="nav-link" href="#traits-and-trait-abbreviations"><span class="header-section-number">4.3</span> Traits and Trait Abbreviations</a></li>
<li><a class="nav-link" href="#qc-trait-values"><span class="header-section-number">4.4</span> QC Trait Values</a></li>
<li><a class="nav-link" href="#post-qc-composite-traits"><span class="header-section-number">4.5</span> Post-QC: composite traits</a></li>
<li><a class="nav-link" href="#season-wide-mean-disease-severity"><span class="header-section-number">4.6</span> Season-wide mean disease severity</a></li>
<li><a class="nav-link" href="#fresh-root-yield-fyld"><span class="header-section-number">4.7</span> Fresh root yield (FYLD)</a></li>
<li><a class="nav-link" href="#check-genotype-to-phenotype-matches"><span class="header-section-number">4.8</span> Check genotype-to-phenotype matches</a></li>
<li><a class="nav-link" href="#check-detect-experimental-designs"><span class="header-section-number">4.9</span> Check / “detect” experimental designs?</a></li>
<li><a class="nav-link" href="#checklist-and-kpi"><span class="header-section-number">4.10</span> Checklist and KPI</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/wolfemd/GenomicSelectionManual/blob/master/02-training_data_QC.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/wolfemd/GenomicSelectionManual/edit/master/02-training_data_QC.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Genomic Prediction and Selection Manual</strong>" was written by Marnin Wolfe. It was last built on 2021-12-05.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
