<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Using this manual | Genomic Prediction and Selection Manual</title>
<meta name="author" content="Marnin Wolfe">
<meta name="description" content="3.1 Workflow notes Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 3 Using this manual | Genomic Prediction and Selection Manual">
<meta property="og:type" content="book">
<meta property="og:url" content="https://wolfemd.github.io/GenomicSelectionManual/using-this-manual.html">
<meta property="og:description" content="3.1 Workflow notes Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Using this manual | Genomic Prediction and Selection Manual">
<meta name="twitter:description" content="3.1 Workflow notes Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Genomic Prediction and Selection Manual</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Preamble</a></li>
<li><a class="" href="outline.html"><span class="header-section-number">2</span> Outline</a></li>
<li><a class="active" href="using-this-manual.html"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="" href="create_project.html"><span class="header-section-number">4</span> Create a project</a></li>
<li><a class="" href="download-training-data.html"><span class="header-section-number">5</span> Download training data</a></li>
<li><a class="" href="prepare-phenotype-data.html"><span class="header-section-number">6</span> Prepare phenotype data</a></li>
<li><a class="" href="prepare-genotypic-data.html"><span class="header-section-number">7</span> Prepare genotypic data</a></li>
<li><a class="" href="preliminary-field-trial-analysis.html"><span class="header-section-number">8</span> Preliminary field trial analysis</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/wolfemd/GenomicSelectionManual">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="using-this-manual" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Using this manual<a class="anchor" aria-label="anchor" href="#using-this-manual"><i class="fas fa-link"></i></a>
</h1>
<div id="workflow-notes" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Workflow notes<a class="anchor" aria-label="anchor" href="#workflow-notes"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>Create a <code>workflowR</code> repository for your genomic prediction analysis, following <a href="create_project.html#create_project">instructions here</a>.</p></li>
<li>
<p>Follow along the following documents as templates and examples:</p>
<ul>
<li><strong>Genomic Selection Manual (THIS BOOK)</strong></li>
<li><strong>GS Process Map (~TO BE LINKED HERE, AND SHOWN THROUGHOUT THE MANUAL~)</strong></li>
<li><strong>GS Checklist (~TO BE LINKED HERE~)</strong></li>
</ul>
</li>
<li><p>Use some variant of these documents and code examples to complete a genomic prediction analysis and develop a report on the results.</p></li>
<li>
<p><strong>Advice and best practices:</strong></p>
<ul>
<li><p>Choose your own data, traits, from cassavabase</p></li>
<li>
<p>Work through what the example code actually does for yourself.</p>
<ul>
<li>Follow-up functions you donâ€™t know by going to the manual.</li>
<li>I will strive to provide references to other tutorials, papers, etc. to give context and help you learn more detail where desired..</li>
</ul>
</li>
<li><p>Inevitably, we will want divergences, alterations, bells-and-whistles on top of the process documented. <strong>SUGGEST</strong> altering and developing your own <strong>process maps</strong> and <strong>checklists</strong> as you go.</p></li>
<li><p>Use a combination of Rmarkdown (<strong>.Rmd</strong>) and Rscripts (<strong>.R</strong>) to document your analysis, as demonstrated.</p></li>
<li><p>Take the time to write commentary throughout. In full sentences, what do you intend to do? How do you interpret the results? What is the next step? Etc.</p></li>
<li><p>Take the time to think through the naming of datasets, files, folders, R objects, etc.</p></li>
<li><p>Use Git version control, made easy with Rstudio.</p></li>
<li><p>Publish your code to GitHub and a report on your results as a webpage using GitHub Pages. I will demonstrate using the package <code>workflowR</code> to manage these aspects.</p></li>
</ul>
</li>
</ul>
</div>
<div id="r-sessions-packages-to-load" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> R sessions, packages to load<a class="anchor" aria-label="anchor" href="#r-sessions-packages-to-load"><i class="fas fa-link"></i></a>
</h2>
<p>I will use the <code>tidyverse</code> and also <code>genomicMateSelectR</code> packages throughout the pipeline.</p>
<p>There are others that may appear.</p>
<p>I recommend, for each pipeline segment, starting with a new R session. Begin each segment, with a step to load these R packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/genomicMateSelectR">genomicMateSelectR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gt.rstudio.com/">gt</a></span><span class="op">)</span> <span class="co"># just for the nice looking tables</span></code></pre></div>
</div>
<div id="software-to-install" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Software to install<a class="anchor" aria-label="anchor" href="#software-to-install"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>R</p></li>
<li><p>Rstudio</p></li>
<li>
<p>Packages</p>
<ul>
<li>tidyverse</li>
<li><code>genomicMateSelectR</code></li>
</ul>
</li>
<li>
<p>Bioinformatics tools</p>
<ul>
<li><code>vcftools</code></li>
<li><code>bcftools</code></li>
</ul>
</li>
</ul>
</div>
<div id="high-performance-and-remote-computing" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> High performance and remote computing<a class="anchor" aria-label="anchor" href="#high-performance-and-remote-computing"><i class="fas fa-link"></i></a>
</h2>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="outline.html"><span class="header-section-number">2</span> Outline</a></div>
<div class="next"><a href="create_project.html"><span class="header-section-number">4</span> Create a project</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#using-this-manual"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="nav-link" href="#workflow-notes"><span class="header-section-number">3.1</span> Workflow notes</a></li>
<li><a class="nav-link" href="#r-sessions-packages-to-load"><span class="header-section-number">3.2</span> R sessions, packages to load</a></li>
<li><a class="nav-link" href="#software-to-install"><span class="header-section-number">3.3</span> Software to install</a></li>
<li><a class="nav-link" href="#high-performance-and-remote-computing"><span class="header-section-number">3.4</span> High performance and remote computing</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/wolfemd/GenomicSelectionManual/blob/master/00-using_this_manual.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/wolfemd/GenomicSelectionManual/edit/master/00-using_this_manual.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Genomic Prediction and Selection Manual</strong>" was written by Marnin Wolfe. It was last built on 2022-01-13.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
